src/nvim/diff.c

```diff
@@ -1782,20 +1771,328 @@ void diff_clear(tabpage_T *tp)
+
+        // get a char* to the beginning of the diff block by incrementing past a
+        // newline character dp->df_lnum[i] times
+        while (*diff_begin[diffbuffers_count] != '\n') {
+          diff_begin[diffbuffers_count]++;
+        }
+        diff_begin[diffbuffers_count]++;
+      }
+
+      // keep track of the length of this diff block to pass it to the linematch
+      // algorithm
+      diff_length[diffbuffers_count] = dp->df_count[i];
+
+      // keep track of the index of the diff buffer we are using here
+      // ( 0 >= i < DB_COUNT )
+      // we need to track it because the order of the indexes used to store the
+      // diff hunk information may not be all in a row e.g.
+
+      // dp->df_count[0] = 5, dp->df_lnum[0] = 10
+      //
+      // dp->df_count[1] = 4, dp->df_lnum[1] = 9  <- if 3 diff buffers were open
+      // and in the middle one, diff mode was disabled with ":diffoff", we are
+      // now no longer interested in using dp->df_lnum/count[1], only
+      // dp->df_lnum/count[0] and [2]. and curtab->tp_diffbuf[1] will be NULL
+      // So we keep track of this with outputmap
+      //
+      // dp->df_count[2] = 6, dp->df_lnum[2] = 11
+
+      outputmap[diffbuffers_count] = i;
+
+      // increment the amount of diff buffers we are passing to the algorithm
+      diffbuffers_count++;
+    }
+  }
+
+  // we will get the output of the linematch algorithm in the format of an array
+  // of integers (*decisions) and the length of that array (decisions_length)
+  int decisions_length = 0;
+  int *decisions = NULL;
+  linematch_nbuffers(diff_begin, diff_length, diffbuffers_count,
+      &decisions_length, &decisions);
+
+  // get the start line number here in each diff buffer, and then increment
+  int line_numbers[DB_COUNT];
+  for (int i = 0; i < DB_COUNT; i++) {
+    if (curtab->tp_diffbuf[i] != NULL) {
+      line_numbers[i] = dp->df_lnum[i];
+      dp->df_count[i] = 0;
+    }
+  }
+  // write the diffs to the current diff block
+  diff_T *dp_s = dp;
+  for (int i = 0; i < decisions_length; i++) {
+    if (i && (decisions[i - 1] != decisions[i])) {
+      // create new sub diff blocks to segment the original diff block which we
+      // further divided by running the linematch algorithm
+      dp_s = diff_alloc_new(curtab, dp_s, dp_s->df_next);
+      dp_s->is_linematched = true;
+      for (int j = 0; j < DB_COUNT; j++) {
+        if (curtab->tp_diffbuf[j] != NULL) {
+          dp_s->df_lnum[j] = line_numbers[j];
+          dp_s->df_count[j] = 0;
+        }
+      }
+    }
+    for (int j = 0; j < diffbuffers_count; j++) {
+      if (decisions[i] & (1 << j)) {
+        // will need to use the map here
+        dp_s->df_count[outputmap[j]]++;
+        line_numbers[outputmap[j]]++;
+      }
+    }
+  }
+  for (int i = 0; i < diffbuffers_count; i++) {
+    clear_diffin(&diffbuffers[i]);
+  }
+  xfree(diffbuffers);
+  xfree(diff_begin);
+  xfree(diff_length);
+  xfree(outputmap);
+  xfree(decisions);
```

Comment:
there is a lot of allocations going on. maybe it is okay for a first feature merge (correctness first), but in case we want to optimize this later on, elimination of allocations will be a low hanging fruit (by using an arena, or static buffers, or both). Could be added as a TODO comment somewhere.